### ISSUE TITLE
 Sharing mode should be part of barrier

### ISSUE THREAD

**Issue by nanokatze** [KhronosGroup/Vulkan-Docs#1721]
Some of the existing implementations with multiple queue families appear to only consider sharing mode during recording of a barrier, but not for memory requirements or anything else. I believe it would make sense to make sharing mode be provided by the user during recording of a barrier.

From API perspective, a possible approach would be as follows:

An extension struct
```
typedef struct Vk??? {
	VkStructureType          sType;
	const void*              pNext;
	uint32_t                 srcQueueFamilyIndexCount;
	const uint32_t*          pSrcQueueFamilyIndices;
	uint32_t                 dstQueueFamilyIndexCount;
	const uint32_t*          pDstQueueFamilyIndices;
} Vk???;
```
would be added for `VkImageMemoryBarrier{,2KHR}`. Both `p{Src,Dst}QueueFamilyIndices` fields would be required to contain index of the queue family the transition is happening on. When recording a barrier that doesn't change sharing, `pSrcQueueFamilyIndices` and `pDstQueueFamilyIndices` would equal each other (exactly or up to permutation, whichever is more sensible). QFOT would be replaced with sharing. For example, to transfer ownership from queue family 0 to 1, the user would first execute a barrier, changing sharing from {0} to {0, 1} and then from {0, 1} to {1}, corresponding to release and acquire barriers in the current API. 

Optionally, `{src,dst}SharingMode` fields could be added, to better match `VkImageCreateInfo` interface.

Assuming the only use of sharing mode is barriers, for all current and future implementations, the current sharing mode and QFOT mechanism could be replaced in favor of a flag at image creation time, to indicate that the image uses the new mechanism.

If some implementations also consider sharing mode for e.g. memory requirements, the extension struct would be used to coerce sharing to subset of sharing specified at creation time. That is, if an image was created shared to families 0, 1, 2, the only possible values (up to permutations) of `p{Src,Dst}QueueFamilyIndices` would be {0}, {1} (collectively EXCLUSIVE if `{src,dst}SharingMode}` fields are added), {0, 1}, {0, 2}, {1, 2} and {0, 1, 2}.

In case of buffers, I'm not aware of what sharing modes and queue family ownership transfers map to, beyond minor optimizations like mapping QFOT barriers to nops (which stems from QFOT not being cancelable), otherwise achievable by moving the barrier at the end of submission, but this suggestion appears to be applicable for buffers as well.

**Comment by martty** [comment]
This new API would allow a more explicit way of describing when a queue acquires a resource without preserving contents by allowing srcQueueFamilyIndices be an empty set when this behavior is required. 

This way it is possible to validate when a queue tries to share in an unintended way (srcQueueFamilyIndices does not match the current sharing set).

**Comment by HansKristian-Work** [comment]
I'm a bit confused which use case this solves. Is there a concrete use case where this solves a problem that cannot be solved with the existing model? To simplify the discussion before considering a new pNext struct, it would be helpful to narrow it down to:

- A specific use case where existing model does not work well enough.
- A desired mechanism that would solve that problem.
- Would that desired mechanism help any vendor extract better performance than is currently possible?

After that, discussing concrete API is appropriate. It's not obvious to me from reading the description if this is an API cleanup, or introducing functionality that cannot be expressed right now.

> Some of the existing implementations with multiple queue families appear to only consider sharing mode during recording of a barrier

IIRC, some implementations consider sharing mode when creating an image, deciding which kind of compression format is used.

**Comment by nanokatze** [comment]
There are no problems that aren't solvable with the current model, but in some niche cases some implementations of the current model will have some unnecessary overhead. Here's a contrived example: on a general queue, 1) copy some image into another image A, 2) we then draw some triangles on top of A, 3) proceed to sample from A (on the general queue) while copying from A into some buffer (on another, perhaps compute-only or transfer-only queue). In the current model, on a sufficiently old radeon running under, say, radv, the transition 1 from `UNDEFINED` to `TRANSFER_DST` will probably initialize that image into a worse-for-performance state if the image was shared to both general and compute-only queue, compared to exclusive, which is what we want at here. It's only at 3 we want the image to be readable from some two queues. In current model, we can only say that we either want sharing always or never at all.

One possible mechanism to solve this is presented in the OP and it would allow extract better performance on at least radeon (which is the only implementation I am aware of that has multiple queue families and is known to benefit from fine layout transitions). But overall, given the depth of the niche this suggestion addresses, let's consider this API cleanup.

> IIRC, some implementations consider sharing mode when creating an image, deciding which kind of compression format is used.

Pure curiosity, but could you name or point to examples of such?

**Comment by nanokatze** [comment]
Thanks for answers. Speaking about radv, I examined it before even thinking about writing this suggestion and the only time sharing was considered at image creation time was when deciding whether to set "disable DCC" and "disable HTILE" surface flags. I got an impression that setting these flags could be avoided (conservatively), although at some memory penalty (space for metadata would still be allocated) and possibly some performance? After all, even images with DCC allocated sometimes need to be used in ways that are ignorant of DCC. But it seems like a huge can of worms so preferably someone competent on that topic would comment on this.
